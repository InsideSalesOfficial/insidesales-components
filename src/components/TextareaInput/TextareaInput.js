import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import _ from 'lodash';

import { colors, typography } from '../styles';

const TextareaInputWrapper = styled.div`
  position: relative;
  width: 100%;
`;

const Textarea = styled.textarea`
  background-color: ${colors.white};
  border: ${(props) => {
    if (props.error) {
      return `2px solid ${colors.red}`;
    } else {
      return `thin solid ${colors.black40}`;
    }
  }};
  border-radius: 4px;
  box-sizing: border-box;
  padding: ${(props) => {
    if (props.error) {
      return '27px 15px 15px';
    } else {
      return '28px 16px 16px';
    }
  }};
  height: 100px;
  min-height: 100px;
  min-width: 100%;
  resize: none;
  text-align: left;
  transition: border-color 0.14s ease-in-out;
  &:hover {
    border-width: ${(props) => {
      if (props.disabled) {
        return '1px';
      }
      return '2px';
    }};
    padding: ${(props) => {
      if (props.disabled) {
        return '28px 16px 16px';
      }
      return '27px 15px 15px';
    }};
  }
  &:focus {
    border: ${(props) => {
      if (props.error) {
        return `2px solid ${colors.red}`;
      } else {
        return `2px solid ${colors.green}`;
      }
    }};
    outline: 0;
    padding: 27px 15px 15px;
    + label {
      color: ${(props) => {
        if (props.error) {
          return colors.red;
        } else {
          return colors.green;
        }
      }};
    }
  }
  &:disabled {
    background-color: ${colors.black05};
    border-color: ${colors.black40};
    + label {
      color: ${colors.black40};
    }
  }
  ${typography.subhead1}
`;

const TextLabel = styled.label`
  color: ${(props) => {
      if (props.error) {
        return colors.red;
      } else {
        return colors.black60;
      }
    }};
  left: 16px;
  position: absolute;
  transform: ${(props) => {
    if (props.open) {
      return 'translateY(10px)';
    } else {
      return 'translateY(19px)';
    }
  }};
  transition: font-size 0.14s ease-in-out, transform 0.14s ease-in-out, color 0.14s ease-in-out;
  ${(props) => {
    if (props.open) {
      return typography.caption
    } else {
      return typography.subhead1
    }
  }}
  line-height: 16px;
`;

const TextareaHelper = styled.div`
  color: ${colors.black40};
  ${typography.caption}
`;

const TextareaError = styled(TextareaHelper)`
  color: ${colors.red};
`;

class TextareaInput extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      focusedOrHasValue: false
    };
  }

  componentDidMount() {
    if (this.props.children) {
      this.setState({
        focusedOrHasValue: true
      });
    }
  }

  focused = () => {
    this.setState({
      focusedOrHasValue: true
    })
  }

  blurred = (e) => {
    if (_.get(e, 'target.value', false)) {
      this.setState({
        focusedOrHasValue: true
      })
    }
    else {
      this.setState({
        focusedOrHasValue: false
      })
    }
  }

  renderHelperText() {
    const { error, helper } = this.props;
    if (error) {
      return (<TextareaError>{error}</TextareaError>);
    }
    return (<TextareaHelper>{helper}</TextareaHelper>);
  }

  render() {
    const { label, name, error, disabled, children } = this.props;
    return (
      <TextareaInputWrapper>
        <Textarea onFocus={this.focused} onBlur={this.blurred} id={name} name={name} error={error} disabled={disabled}>
          {children}
        </Textarea>
        <TextLabel open={this.state.focusedOrHasValue} htmlFor={name} error={error}>{label}</TextLabel>
        {this.renderHelperText()}
      </TextareaInputWrapper>
    );
  }
}

TextareaInput.defaultProps = {
  name: 'Name',
  label: 'Label'
};

TextareaInput.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  helper: PropTypes.string,
  error: PropTypes.string,
  disabled: PropTypes.bool,
  children: PropTypes.string
};

export default TextareaInput;
