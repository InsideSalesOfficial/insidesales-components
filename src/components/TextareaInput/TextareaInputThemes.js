import { colors, ifThemeInPropsIsPresentUse } from "../styles/colors";

const originalShared = {
  disabledBorderColor: colors.black20,
  disabledBackgroundColor: colors.black05,
  focusedBorderColor: colors.green,
  errorBorderColor: colors.red,
  disabledColor: colors.black40,
  placeholderColor: colors.black40,
  labelErrorColor: colors.red,
  labelFocusedColor: colors.green
};

export const darkTheme = {
  ...originalShared,
  background: colors.darkBlueC,
  valueColor: colors.white90,
  labelColor: colors.white60,
  borderColor: colors.white40,
  helperColor: colors.white60,
  textAreaInputDarkTheme: true
};

export const defaultTheme = {
  ...originalShared,
  background: colors.white,
  valueColor: colors.black,
  labelColor: colors.black60,
  borderColor: colors.black40,
  helperColor: colors.black40,
  textAreaInputWhiteTheme: true
};

export default { darkTheme, defaultTheme };

function sharedGlobalTheme(theme) {
  return {
    focusedBorderColor: theme.brand01,
    errorBorderColor: theme.warning04,
    placeholderColor: theme.white40,
    labelFocusedColor: theme.brand01,
    labelErrorColor: theme.warning04
  };
}

export function generateFlatTheme(theme) {
  return {
    ...sharedGlobalTheme(theme),
    background: theme.primary01,
    disabledBackgroundColor: theme.primary01,
    valueColor: theme.white90,
    labelColor: theme.white60,
    borderColor: theme.white40,
    helperColor: theme.white60,
    disabledBorderColor: theme.white10
  };
}

export function generateAlternateTheme(theme) {
  return {
    ...sharedGlobalTheme(theme),
    background: theme.primary05,
    disabledBackgroundColor: theme.primary05,
    valueColor: theme.white90,
    labelColor: theme.white60,
    helperColor: theme.white60,
    borderColor: theme.transparent,
    disabledBorderColor: theme.transparent
  };
}

function resolveToModifiedTheme({ theme }) {
  if (theme.textAreaInputWhiteTheme) return generateFlatTheme(theme);
  return generateAlternateTheme(theme);
}

export function themeToThemeResolver({ theme, key }) {
  if (!ifThemeInPropsIsPresentUse({ props: { theme }, value: true }))
    return theme[key];
  return resolveToModifiedTheme({ theme })[key];
}
