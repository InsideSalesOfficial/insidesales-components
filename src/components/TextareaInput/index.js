export { default } from './TextareaInput';
export { darkTheme } from './TextareaInputThemes';