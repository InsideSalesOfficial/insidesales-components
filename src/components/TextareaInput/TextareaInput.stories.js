import React from 'react';
import {
  storiesOf
} from '@storybook/react';

import TextareaInput from './TextareaInput';

storiesOf('Form', module)
  .addWithChapters(
    'TextareaInput',
    {
      info: `
        Usage

        ~~~
        import React from 'react';
        import {TextareaInput} from 'insidesales-components';
        ~~~
      `,
      chapters: [
        {
          sections: [
            {
              title: 'Example: textarea default empty',
              sectionFn: () => (
                <TextareaInput
                  label="Label"
                  name="first"
                />
              )
            },
            {
              title: 'Example: textarea with existing text',
              sectionFn: () => (
                <TextareaInput
                  label="Label"
                  name="firstz"
                >
                  This text was hardcoded into stories. The structure of this component follows how a `textarea` should look.
                </TextareaInput>
              )
            },
            {
              title: 'Example: textarea with helper text',
              sectionFn: () => (
                <TextareaInput
                  label="Labely"
                  helper="Helper text."
                  error=""
                  name="second"
                />
              )
            },
            {
              title: 'Example: textarea with error text',
              sectionFn: () => (
                <TextareaInput
                  label="Label"
                  helper="Helper text."
                  error="Errors will override helper text."
                  name="third"
                />
              )
            },
            {
              title: 'Example: textarea disabled',
              sectionFn: () => (
                <TextareaInput
                  label="Label"
                  helper="Helper text."
                  disabled
                  name="fourth"
                />
              )
            },
            {
              title: 'Example: textarea disabled with text',
              sectionFn: () => (
                <TextareaInput
                  label="Label"
                  helper="Helper text."
                  disabled
                  name="fifth">
                    this is some example text
                </TextareaInput>
              )
            },
            {
              title: 'Example: textarea closed without value',
              sectionFn: () => (
                <TextareaInput
                  label="Label"
                  helper="Helper text."
                  name="sixth"
                  collapsed>
                </TextareaInput>
              )
            },
            {
              title: 'Example: textarea closed without value and has an error',
              sectionFn: () => (
                <TextareaInput
                  label="Label"
                  helper="Helper text."
                  error={"This field is required."}
                  name="sixth"
                  collapsed>
                </TextareaInput>
              )
            },
          ]
        }
      ]
    }
  );