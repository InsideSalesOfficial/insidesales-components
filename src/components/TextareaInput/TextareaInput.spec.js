import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import TextareaInput from './TextareaInput';

describe('TextareaInput', () => {
  
  it('componentDidMount should set focusedOrHasValue on state if there are any children', () => {
    const initialValue = true;
    const wrapper = mount(<TextareaInput name="test">Has text here</TextareaInput>);
    
    expect(wrapper.state().focusedOrHasValue).toEqual(initialValue);

  });
});

describe('TextareaInput Snapshots', () => {
  test('shows component with a label', () => {
    const tree = renderer.create(<TextareaInput name="Name" label="test label" />);
    expect(tree).toMatchSnapshot();
  });
});
