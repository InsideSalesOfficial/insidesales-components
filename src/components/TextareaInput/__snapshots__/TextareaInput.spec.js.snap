// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaInput Snapshots shows component with a label 1`] = `
<div
  className="sc-bdVaJa jGgciL"
>
  <textarea
    className="sc-bwzfXH ddUNhN"
    disabled={undefined}
    id="Name"
    name="Name"
    onBlur={[Function]}
    onFocus={[Function]}
  />
  <label
    className="sc-htpNat kahHcM"
    htmlFor="Name"
    open={false}
  >
    test label
  </label>
  <div
    className="sc-bxivhb dWaEaW"
  />
</div>
`;
